### ISSUE TITLE
effect worker thread on loss in word2vec sgns

### ISSUE THREAD

**Issue by grv1207** [RaRe-Technologies/gensim#2222]
Im trying to train a gensim(3.6.0) sgns model and in the process, I measure the loss during which I'm calculating as

loss = model.running_training_loss / model.corpus_count, 

however, I noticed that if I change my worker thread I get different losses keeping all other parameters same. Especially if I keep my worker thread a 1 I get a really high loss and If I increase the number of threads I get less loss. An instance

thread  loss
worker=1  20.40519721
worker=10   2.714875407
worker=16  1.239528453
what could be the reason behind it ..?

**Comment by menshikh-iv** [comment]
Hello @grv1207, can you provide a full code example, please?

CC: @gojomo

**Comment by gojomo** [comment]
As I mentioned in my [comment on SO](https://stackoverflow.com/questions/52743468/effect-of-increase-worker-thread-in-gensim-word2vec/52745921#comment92453092_52745921) (emphasis added):

It's possible there's something amiss with your not-shown-here training code, or the collecting/reporting of the varying loss values. But perhaps also there's a bug in gensim 3.6.0. If you can make a **self-contained code/data example showing extreme changes in reported-loss with different worker values**, you should report it as a bug in the gensim github issue tracker.

**Comment by grv1207** [comment]
Here is the snippet that I wrote
import numpy as np
from gensim.models import Word2Vec
from gensim.models.callbacks import CallbackAny2Vec
from gensim.parsing.preprocessing import strip_non_alphanum
from nltk.corpus import stopwords

from commonFunctions import MMOLogger

stopWords = set(stopwords.words('english'))
#mylogger = MMOLogger().getLogger(__name__, 'status.log')


def getline(filepath):
    with open(filepath,'r') as fout:
        for line in fout:
            line_list = strip_non_alphanum(line).split()

            yield([token for token in line_list if token not in stopWords])


class my_callback_each_epoch(CallbackAny2Vec):


    def __init__(self,word2vec_model,my_min_count):
        self.best_loss = np.inf
        self.five_score_list = []
        self.epoch_dict = {}
        self.epoch = 0
        self.word2vec_model = word2vec_model
        self.my_min_count = my_min_count
        self.stop_training = False


    def on_epoch_begin(self, model):
        #print("Epoch #{} start".format(self.epoch))
        if self.stop_training:
            print(self.epoch_dict)
            exit()

    def on_epoch_end(self, model):
        #print("Epoch #{} end".format(self.epoch))
        loss = model.running_training_loss / model.corpus_count
        # print('error : {} '.format(loss))
        print(' present error : {}, best error {} '.format(loss, self.best_loss))
        self.epoch_dict[self.epoch] = loss
        self.epoch += 1
        if loss < self.best_loss:
            self.best_loss = loss
            self.five_score_list = []

            if not os.path.exists(str(self.my_min_count)):
                os.makedirs(str(self.my_min_count))

            print('saving in {}.bin'.format( self.word2vec_model))
            model.save(str(self.my_min_count) + '/' +  self.word2vec_model + '.bin')
            print(' saving in {}.vec '.format( self.word2vec_model))
            model.wv.save(str(self.my_min_count) + '/' +  self.word2vec_model + '.vec')
            model.wv.save_word2vec_format(str(self.my_min_count) + '/' +  self.word2vec_model + '.txt', binary=False)

        else:
            #self.epoch_dict[self.epoch] = loss
            self.five_score_list.append(loss)

        if len(self.five_score_list) ==3:

            self.stop_training = True

def main(file_path, word2vec_model, my_min_count,dimension, negative_size,lr,context_window,worker_thread):
    my_callback = my_callback_each_epoch(word2vec_model, my_min_count)

    model = Word2Vec(sentences= [token for token in getline(file_path)], sg=1,alpha= lr,min_count=my_min_count,size=dimension,sample=0.00001,\
                     negative=negative_size,workers=worker_thread,window=context_window,compute_loss=True,callbacks=[my_callback],iter=100,seed=1)

**Comment by gojomo** [comment]
This code doesn't show the full set of tests/parameters that report different losses for different `worker` counts. It would even be better if you could demonstrate a problem without the use of any `callbacks` that do anything other than print losses.

**Comment by menshikh-iv** [comment]
I close an issue because of the lack of information. Feel free to re-open when an example is ready
